3.13/specializer.md

```diff
@@ -0,0 +1,289 @@
+in the type system, for example, dictionary versions or type verions.
+
+These rules follow deductive steps. Type information is propagated forwards.
+For example, in the IR:
+```
+stack0  := INIT_FAST a1
+stack1  := INIT_FAST b1
+# types: (a1: unknown, b1: unknown)
+ir_inst(GUARD_BOTH_INT)
+# types: (a1: int, b1: int)
+x       := BINARY_OP_ADD_INT(c1, BINARY_OP_ADD_INT(stack0, stack1))
+# types: (a1: int, b1: int, x: int)
+null    := CALL(thing1)
+```
+
+The key part is that type and constant propagation is done automatically,
+without having to handwrite these deductive rules ourselves. Types are
+expressed as part of the interpreter DSL:
+
+```
+// Typed inputs mean the inputs are these types after the operation.
+guard op(_GUARD_BOTH_INT, (left: ~(PYINT_TYPE), right: ~(PYINT_TYPE) -- left, right)) {
+}
+
+// Typed outputs means the types are these types after the operation.
+pure op(_BINARY_OP_ADD_INT, (left, right -- res: ~(PYINT_TYPE))) {
+}
+```
+
+The abstract interpreter's then automatically generates all of the
+cases. With types, we can then eliminate type guards that are not
+required, thus shrinking the code we eventually pass to the JIT
+compiler.
+
+### Constant propagation and evaluation
+
+Like with type propagation, the abstract interpreter automatically
+evaluates the bodies of pure instructions with their constant values.
+We get this for free.
+
+### Guard elimination
+We eliminate guards when we see they are redundant using constant
+or type information.
+
+### True function inlining
+
+CPython 3.11 obtained frame inlining. This eliminated most of the
+overhead of function calls. However, "fake frames"
+were still needed. These are the instructions `_PUSH_FRAME` and `_POP_FRAME`.
+A similar idea was introduced in Cinder (Instagram's fork of CPython).
+
+In CPython 3.13, we plan for true function inlining -- that is, no
+frame creation at all in the happy case, not even fake frames.
+
+The uops tracer already projects traces through Python function boundaries.
+The goal is to eliminate `_PUSH_FRAME` and `_POP_FRAME`. Doing this is easy,
+but doing this while maintaining CPython semantics is very hard.
+
+The goal is to change this trace of instructions:
+```
+ADD_TO_TRACE(_CHECK_PEP_523, 1, 0)
+ADD_TO_TRACE(_CHECK_FUNCTION_EXACT_ARGS, 1, 4828)
+ADD_TO_TRACE(_CHECK_STACK_SPACE, 1, 0)
+# Frame creation
+ADD_TO_TRACE(_INIT_CALL_PY_EXACT_ARGS, 1, 0)
+ADD_TO_TRACE(_SAVE_RETURN_OFFSET, 4, 0)
+ADD_TO_TRACE(_PUSH_FRAME, 1, 0)
+ADD_TO_TRACE(_SET_IP, 0, 0)
+ADD_TO_TRACE(_CHECK_VALIDITY, 0, 0)
+ADD_TO_TRACE(_RESUME_CHECK, 0, 0)
+...
+# Frame removal
+ADD_TO_TRACE(_POP_FRAME, 0, 0)
+```
+to this:
+```
+_CHECK_PEP_523(1, 12, 0)
+_CHECK_FUNCTION_EXACT_ARGS(1, 12, 4828)
+_CHECK_STACK_SPACE(1, 12, 0)
+_SETUP_TIER2_FRAME(128, 0, 0)
+ # Minimal bookkeeping, no frame creation
+_PRE_INLINE(7, 0, 39)
+_SET_FRAME_NAMES(0, 0, 93831592813424)
+...
+_POST_INLINE(4, 0, 18446744073709551615)
+_SET_FRAME_NAMES(0, 0, 93831592813424)
+```
```

Comment:
You reduce the micro-op count from 10 to 8, but with other, more general optimizations we can reduce the number of micro-ops to 7.

Assuming the ability to support global deoptimization, and noting that most functions are accessed through global variables, we can assume that the function is a constant.
We can also use  global deoptimization to eliminate PEP_523 checks.

With that in mind, the above code becomes:
```
ADD_TO_TRACE(_CHECK_STACK_SPACE, 1, 0)
# Frame creation
ADD_TO_TRACE(_INIT_CALL_PY_EXACT_ARGS, 1, 0)
ADD_TO_TRACE(_SAVE_RETURN_OFFSET, 4, 0)
ADD_TO_TRACE(_PUSH_FRAME, 1, 0)
ADD_TO_TRACE(_SET_IP, 0, 0)
ADD_TO_TRACE(_RESUME_CHECK, 0, 0)
...
# Frame removal
ADD_TO_TRACE(_POP_FRAME, 0, 0)
```

This not a criticism of the technique, just to show that there are other optimizations we should be doing first.